src/main/java/apoc/merge/Merge.java

```diff
@@ -22,49 +23,88 @@
+        }
+
+        validateKeys(config.keySet(), Lists.newArrayList("onCreateProps", "onMatchProps"));
+
+        String labels = labelNames.stream().map(s -> Util.quote(s)).collect(Collectors.joining(":"));
+
+        Map<String, Object> onCreateProps = (Map<String, Object>) config.getOrDefault("onCreateProps", Collections.emptyMap());
+        Map<String, Object> onMatchProps = (Map<String, Object>) config.getOrDefault("onMatchProps", Collections.emptyMap());
+
+        Map<String, Object> params = Util.map("identProps", identProps, "onCreateProps", onCreateProps, "onMatchProps", onMatchProps);
+        String identPropsString = buildIdentPropsString(identProps);
+
+        final String cypher = "MERGE (n:" + labels + "{" + identPropsString + "}) ON CREATE SET n += $onCreateProps ON MATCH SET n += $onMatchProps RETURN n";
+        return db.execute(cypher, params ).columnAs("n").stream().map(node -> new NodeResult((Node) node));
     }
 
     @Procedure(mode = Mode.WRITE)
-    @Description("apoc.merge.relationship(startNode, relType,  {key:value, ...}, {key:value, ...}, endNode) - merge relationship with dynamic type")
+    @Description("apoc.merge.relationship(startNode, relType,  identProps:{key:value, ...}, {key:value, ...}, endNode) - merge relationship with dynamic type")
     public Stream<RelationshipResult> relationship(@Name("startNode") Node startNode, @Name("relationshipType") String relType,
                                                    @Name("identProps") Map<String, Object> identProps, @Name("props") Map<String, Object> props, @Name("endNode") Node endNode) {
         String identPropsString = buildIdentPropsString(identProps);
 
         Map<String, Object> params = new HashMap<>();
-        params.put("props", Util.merge(identProps, props));
+        params.put("identProps", identProps);
+        params.put("props", props != null ? props : Collections.emptyMap());
         params.put("startNode", startNode);
         params.put("endNode", endNode);
 
-        final String cypher = "WITH $startNode as startNode, $endNode as endNode MERGE (startNode)-[r:"+ wrapInBacktics(relType) +"{"+identPropsString+"}]->(endNode) ON CREATE SET r+= $props RETURN r";
+        final String cypher = "WITH $startNode as startNode, $endNode as endNode MERGE (startNode)-[r:"+ Util.quote(relType) +"{"+identPropsString+"}]->(endNode) ON CREATE SET r+= $props RETURN r";
         Relationship rel = Iterators.single(db.execute(cypher, params ).columnAs("r"));
         return Stream.of(new RelationshipResult(rel));
     }
 
+    @Procedure(value = "apoc.merge.relationship.eager", mode = Mode.WRITE, eager = true)
+    @Description("apoc.merge.relationship.eager(startNode, relType,  identProps:{key:value, ...}, config:{onCreateProps:{key:value, ...}, onMatchProps:{key:value, ...}}, endNode) - merge relationship with dynamic type, with support for setting properties ON CREATE or ON MATCH")
```

Comment:
I wouldn't put the extra props into config but into optional parameters which default to empty maps.